✨  feat : 예약 & 결제 통합 생성 API 구현, PortOne API 연동 설정 추가

j-hann opened the pull request:
**✨ 예약 & 결제 통합 생성 API 구현**
</br>

🟪 `Payment`, `Reservation`, `Member` Entity 추가 & 수정
   - Member Entity
      - `Reservation`, `PaymentHistory` Entity 단방향 설정을 위해 필드 삭제
   - Reservation Entity
      - `cancelReason(예약 취소 사유)`, `expiredAt(예약 만료 일시)` 필드 추가
      - 예약 만료 상태 변경 및 시점 기록을 위한 `expireReservation` 메서드 추가
   - Payment Entity
     - `merchantUid(주문 고유 번호)`, `paymentMethodType(결제 수단 타입)`, `paidAt(결제 승인일)`, `expiredAt(결제 만료일)` 필드 추가
     - 결제 만료 상태 변경 및 시점 기록을 위한 `expirePayment` 메서드 추가
</br>

🟪 `PaymentStatus`, `ReservationStatus` 상수 추가 
   - `PaymentStatus` : `PAYMENT_EXPIRED (결제 만료)`
   - `ReservationStatus` : `REQUESTED (예약 요청)`, `PAYMENT_PENDING (결제 대기)`
</br>

🟪 예약 & 결제 관련 `Controller` 생성
  - ReservationContoller
     - `createReservationAndPayment` API 생성
     - 예약과 결제를 하나의 요청으로 처리
     - RequestDto : `ReservationAndPaymentRequestDto`
     - ResposeDto : `ReservationAndPaymentResponseDto`
</br>

🟪 예약 & 결제 관련 `DTO` 생성
  - `ReservationCreateRequestDto`, `PaymentCreateRequestDto`를 포함한 `ReservationAndPaymentRequestDto` Wapper DTO 생성
  - `ReservationResponseDto`, `PaymentResponseDto`를 포함한 `ReservationAndPaymentResponseDto` Wapper DTO 생성
</br>

🟪 예약 & 결제 관련 `Service` 생성
  - ReservationApplicationService
     - `createReservationAndPayment` 메서드 구현
     - 예약 및 결제 생성을 하나의 트랜잭션으로 묶어서 처리
     - `createReservation` 메서드 호출하여 예약 생성
     - `createPayment` 메서드 호출하여 결제 생성
     - `결제 검증`, `결제 내역 생성`, `결제 실패처리`, `만료 처리` 등 로직은 추후 추가 예정
     </br>
  - ReservationService
     - `createReservation` 메서드 구현
     - 롤 검증하여 LOCAL_CREATOR는 예약 불가 처리
     - 예약 생성 시, 상태는 `REQUESTED`
     - 결제 생성 시, 사용 할 Reservation 객체 반환
     - `동시성 제어`, `예외처리`, `검증` 등 로직은 추후 추가 예정
      </br>
  - PaymentService
     - `createPayment` 메서드 구현
     - 예약 상태 `PAYMENT_PENDING`으로 업데이트
     - `merchantUid` 주문 고유 번호 생성하는 `generateUniqueMerchantUid` 메서드 분리
       - `res-{yyyyMMdd}-{reservationId}-{랜덤10자리}` 형식으로 조합
       - 현재날짜 + 예약 고유 식별자 + UUID 기반 랜덤 문자열 조합하여 생성
       - `do-while`문 사용하여 merchantUid 중복 여부를 체크하여 충돌 방지
     - 결제 생성 시, `PAYMENT_PENDING` 상태
     - 예약 & 결제 통합 메서드에서 사용 할 Payment 객체 반환

</br>

🟪 ErrorCode 추가
</br>

🟪 Iamport(PortOne) API 설정 추가
   - `build.gradle` 의존성 추가
      - `implementation 'com.github.iamport:iamport-rest-client-java:0.2.23'` 
      - `maven { url 'https://jitpack.io' }` 
   - IamportConfig 생성하여 환경변수 설정
</br>

🟪 추후 프론트에서 PortOne API 연동 후, 테스트 예정

j-hann committed:
✨  feat : 예약 & 결제 통합 생성 API 구현, PortOne API 연동 설정 추가

- 예약 & 결제 통합 생성 API 구현, - Payment, Reservation, Member Entity 수정, - 예약 & 결제 관련 Dto 생성, - Iamport 설정 추가

j-hann committed:
🔀  branch : develop merge

develop 브랜치 merge

j-hann committed:
🔀  branch : develop branch merge

- develop branch merge

sooyeoneo commented on src/main/java/com/salayo/locallifebackend/domain/payment/dto/PaymentCreateRequestDto.java:
```diff
@@ -25,4 +23,6 @@ public class PaymentCreateRequestDto {
 	@NotBlank(message = "결제 카드 정보는 필수값입니다.")
 	private String paymentCard;
 
+	private String paymentMethodType;
```
궁금해서 질문 : 결제 수단은 null 허용인거죠? 🤓

j-hann replied:
넵! 결제 완료 전에도 결제 상태 추척 목적으로 entity 생성하는 구조라서, 둘다 결제 완료 후에 PG사에서 응답 받을 수 있는 값들이라 일단 null 허용으로 해놨어요!

sooyeoneo commented on src/main/java/com/salayo/locallifebackend/global/config/IamportConfig.java:
```diff
@@ -0,0 +1,28 @@
+@Configuration
+public class IamportConfig {
+
+	@Value("{IAMPORT_API_KEY}")
```
@Value("{IAMPORT_API_KEY}") 부분에$가 누락된 것으로 보입니다! 

 특별한 이유가 없다면 $를 추가하는 것은 어떨까요? 😚

```
@Value("${IAMPORT_API_KEY}")

```
아래도 동일합니다!
```
@Value("${IAMPORT_API_SECRET_KEY}")

j-hann replied:
네넵 수정하겠습니다!👍

sooyeoneo commented on src/main/java/com/salayo/locallifebackend/domain/payment/service/PaymentService.java:
```diff
@@ -1,10 +1,69 @@
+		String merchantUid;
+
+		do {
+			merchantUid = "res-" + reservationId + "-" + UUID.randomUUID().toString().replace("-", "").substring(0, 10);
```
현재 `generateUniqueMerchantUid()`에서 UUID를 사용하여 고유성을 확보한 것으로 보여지는데요!
`substring(0, 10)`으로 길이를 줄이면 충돌 확률이 증가할 수 있을 것 같습니다! 🤓

제안 : 안정성을 위해 전체 UUID를 사용하거나 더 긴 문자열을 사용하는 것은 어떨까요? 💓
```
private String generateUniqueMerchantUid(Long reservationId) {

		String merchantUid;

		do {
			// 현재는 10자리로 잘라 사용하지만, 고유성 강화를 위해 전체 UUID를 사용 또는 더 긴 문자열을 고려
			// ex. UUID.randomUUID().toString().replace("-", "") (32자리)
			merchantUid = "res-" + reservationId + "-" + UUID.randomUUID().toString().replace("-", "").substring(0, 10);
		} while (paymentRepository.existsByMerchantUid(merchantUid));

		return merchantUid;
	}

j-hann replied:
제안 감사합니다 수연님!👍🩵 고민해봤는데 

```
do {
	 String date = LocalDate.now().format(DateTimeFormatter.BASIC_ISO_DATE);
	 String randomStr = UUID.randomUUID().toString().replace("-", "").substring(0, 10);
	 merchantUid = "res-" + date + "-" + reservationId + "-" + randomStr ;
} while (paymentRepository.existsByMerchantUid(merchantUid));
```

추후에 해당 필드가 주문번호 역할로 프론트에서 사용자한테 노출되는 경우도 생각을 해본다면, 
`현재 날짜` + `예약 고유 식별자` + `랜덤 10자리` 조합이면 중복으로 인한 충돌 위험성 더 낮을 것 같은데 어떠실까요!

sooyeoneo replied:
좋습니다!!! 짱!!! 🙌

sooyeoneo reviewed (approved):
고생하셨습니다!! 짱짱~~~🩷

j-hann committed:
✨  feat : PaymentService 주문 고유 번호 생성 시, 날짜 포함 로직 추가 & IamportConfig 환경변수 주입 방식 수정

- 주문 고유 번호 생성 시, res-날짜-예약ID-랜덤10자리 조합으로 충돌 가능성 낮춤, - @Value로 환경변수 주입 시, 값 제대로 주입받기 위해 $ 추가

j-hann committed:
Revert "✨  feat : PaymentService 주문 고유 번호 생성 시, 날짜 포함 로직 추가 & IamportConfig 환경변수 주입 방식 수정"

This reverts commit f1baca797c0c69c1f5194412f0a2b71326ca097a.

jiyoon0000 commented:
👍😎🩷🫢

j-hann committed:
✨  feat : PaymentService 주문 고유 번호 생성 시, 날짜 포함 로직 추가 & IamportConfig 환경변수 주입 방식 수정

- 주문 고유 번호 생성 시, `res-날짜-예약ID-랜덤10자리` 조합으로 충돌 가능성 낮춤
- @Value로 환경변수 주입 시, 값 정상 주입되도록 $ 추가
